### PR TITLE
Fix: Issue #2307

### LINKED ISSUE #2307: use pydantic tagged unions to more quickly validate the result document
In our result document JSON document, we use unions in a few places, such as `freeze.features.Feature`:
https://github.com/mandiant/capa/blob/c409b2b7ed7f5f2a9c3b9efc9e75dd5779205576/capa/features/freeze/features.py#L351

image

I've learned that to validate incoming data against this union, pydantic will loop through each case and then use the best match. because there are lots of cases, pydantic does a lot of work.

we should be able to improve performance by using tagged unions. there's a bit of documentation here: 
https://docs.pydantic.dev/latest/concepts/unions/#discriminated-unions

and a PyCon US '24 talk here: https://www.youtube.com/watch?v=Qvj5e9xtaSE

we should investigate using tagged unions and see if it improves performance of loading serialized result documents. granted, this isn't a common operation, but the data model would be more correct regardless.

### PR DESCRIPTION
**Pull request by harshit-wadhwani**
closes #2307

### Checklist

- [x] No CHANGELOG update needed
- [x] No new tests needed
- [x] No documentation update needed

### CODE REVIEW

**Comment by harshit-wadhwani** [comment]
`Annotated` was introduced in Python 3.9.

**Comment by williballenthin** [comment]
we'll remove py3.8 in the next couple days and then we can retry the CI runs. thanks @harshit-wadhwani !

**Comment by williballenthin** [comment]
those mypy errors are outside of the code that @harshit-wadhwani has touched, though i agree it would be great if they can fix the errors.

it seems to be due to stricter checking that mypy can now do: mypy now knows that only some strings can be put into the `type` field, rather  than  any string, so it complains when a string is assigned there. to fix, we should assert that the string that we want to put there is one of the acceptable strings. something like `assert type_name in {"os", "arch", ...}`. except i'm not sure we have this list yet today, so it may be a new global constant to add and maintain.

**Comment by harshit-wadhwani** [comment]
I will take a look into those errors.

**Comment by mr-tz** [comment]
@harshit-wadhwani have you had a chance to take a look here, yet?

**Comment by harshit-wadhwani** [comment]
I ran pre-commit with dev container and it passes everything, but not on the github actions, any suggestions ?

File: capa/ida/plugin/model.py [hunk 1]
```diff
@@ -529,12 +548,40 @@ def render_capa_doc(self, doc: rd.ResultDocument, by_function: bool):
...
+                "number",
+                "bytes",
+                "offset",
+                "mnemonic",
+                "operand number",
+                "operand offset",
+                "basic block",
+            ]
+        ]
+        key: FeatureType = feature.type
```
Comment (mr-tz) [inline capa/ida/plugin/model.py:584]: use a string here (just for rendering)? (and same in vverbose?)

**Comment by mr-tz** [comment]
@williballenthin could you take a look if this is what you had in mind to close out the issue, please?

**Review by williballenthin — APPROVED** [review]
lgtm!

**Comment by mr-tz** [comment]
Thanks, @harshit-wadhwani!

**Pull request merged**